### ISSUE TITLE
Container restart policies may fail on linked containers after engine restart

### ISSUE THREAD

**Issue by BrianBland** [docker/docker#17118]
Description of problem: When running several linked containers with `--restart=always`, some containers may not come back up after an engine restart.

`docker version`:

```
Client:
 Version:      1.9.0-rc1
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   9291a0e
 Built:        Wed Oct 14 03:04:22 UTC 2015
 OS/Arch:      linux/amd64

Server:
 Version:      1.9.0-rc1
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   9291a0e
 Built:        Wed Oct 14 03:04:22 UTC 2015
 OS/Arch:      linux/amd64
```

Experienced on at least 1.6.2, 1.8.2, 1.8.3 as well.

`docker info`:

```
Containers: 37
Images: 409
Engine Version: 1.9.0-rc1
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 485
 Dirperm1 Supported: false
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 3.13.0-39-generic
Operating System: Ubuntu 14.04.1 LTS
CPUs: 2
Total Memory: 1.955 GiB
Name: vagrant-ubuntu-trusty-64
ID: 3YCN:IHBG:FR4D:TDIK:EL4I:YIIY:YUSE:BX73:DTYO:JQOO:A7HL:ZNK2
Username: brianbland
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
```

`uname -a`:

```
Linux vagrant-ubuntu-trusty-64 3.13.0-39-generic #66-Ubuntu SMP Tue Oct 28 13:30:27 UTC 2014 x86_64 x86_64 x86_64 GNU/Linux
```

Environment details (AWS, VirtualBox, physical, etc.):
Ubuntu 14.04 VirtualBox (vagrant)

How reproducible:
Seems to happen ~1/3 of the time when running Docker Trusted Registry (1.0 - 1.3.3), but I have been unable to find a simpler container configuration that commonly reproduces this issue

Steps to Reproduce:
1. Install/run Docker Trusted Registry: `docker run docker/trusted-registry:1.3.3 install | bash`
2. Restart the engine: `service docker restart`

Actual Results:

```
$ docker ps -q | wc -l
6
$ sudo service docker restart
docker stop/waiting
docker start/running, process 31683
$ docker ps -q | wc -l
6
$ sudo service docker restart
docker stop/waiting
docker start/running, process 32127
$ docker ps -q | wc -l
4
```

The two containers that did not restart were linked to other containers in this stack.

Expected Results:
All of the containers running before the engine restart are once again running.

Additional info:
Relevant engine logs:

```
Waiting for /var/run/docker.sock
INFO[0000] API listen on /var/run/docker.sock           
INFO[0000] [graphdriver] using prior storage driver "aufs" 
INFO[0000] Firewalld running: false                     
WARN[0000] Your kernel does not support swap memory limit. 
/var/run/docker.sock is up
INFO[0000] Loading containers: start.                   
...............................ERRO[0000] Failed to load container b68e5d0a08034373459bd7ab7a88e67773887b778a912ae5eda7f0f145e1a5f2: open /var/lib/docker/containers/b68e5d0a08034373459bd7ab7a88e67773887b778a912ae5eda7f0f145e1a5f2/config.json: no such file or directory 
...........................................................WARN[0000] Could not connect container 7aca6e27e3950064df00fb2dfb630360e69457410426f87693b1d42fb4e7876a : nosuchcontainer: no such id: 28dcf8584e35003d559ff1166cae820e5e8a43400adb7057e88d10f5f4c2c640 
ERRO[0000] failed to umount /var/lib/docker/containers/7aca6e27e3950064df00fb2dfb630360e69457410426f87693b1d42fb4e7876a/shm: invalid argument 
ERRO[0000] failed to umount /var/lib/docker/containers/7aca6e27e3950064df00fb2dfb630360e69457410426f87693b1d42fb4e7876a/mqueue: invalid argument 
ERRO[0000] 7aca6e27e3950064df00fb2dfb630360e69457410426f87693b1d42fb4e7876a: Failed to umount ipc filesystems: failed to cleanup ipc mounts:
invalid argument
invalid argument 
ERRO[0000] Failed to start container 7aca6e27e3950064df00fb2dfb630360e69457410426f87693b1d42fb4e7876a: nosuchcontainer: no such id: 28dcf8584e35003d559ff1166cae820e5e8a43400adb7057e88d10f5f4c2c640 
ERRO[0000] failed to umount /var/lib/docker/containers/7191c0055251105be6c21645badc9848124a0caab5e4b7200a8fa914d1ec7929/shm: invalid argument 
ERRO[0000] failed to umount /var/lib/docker/containers/7191c0055251105be6c21645badc9848124a0caab5e4b7200a8fa914d1ec7929/mqueue: invalid argument 
ERRO[0000] 7191c0055251105be6c21645badc9848124a0caab5e4b7200a8fa914d1ec7929: Failed to umount ipc filesystems: failed to cleanup ipc mounts:
invalid argument
invalid argument 
ERRO[0000] Failed to start container 7191c0055251105be6c21645badc9848124a0caab5e4b7200a8fa914d1ec7929: linknotrunning: Cannot link to a non running container: /docker_trusted_registry_admin_server AS /docker_trusted_registry_load_balancer/admin_server 

INFO[0000] Loading containers: done.                    
INFO[0000] Daemon has completed initialization          
INFO[0000] Docker daemon                                 commit=9291a0e execdriver=native-0.2 graphdriver=aufs version=1.9.0-rc1
```

**Comment by cpuguy83** [comment]
Is this a regression from 1.8.x?

**Comment by BrianBland** [comment]
No, this behavior was seen in 1.8.2 and 1.6.2 as well.

**Comment by calavera** [comment]
I'm pretty sure there have never been guarantees to prevent linked containers from starting before their parents. However, before 1.7, containers started in order, which might have given the false perspective of starting in the right order. We change the restarting logic in 1.7 to start all containers in parallel, allowing plugin containers to boot without being blocked by those containers that depend on the plugins. This change might have made this bug more prominent.

All in all, I don't think we should add this to 1.9, or maybe we should not fix it at all. AFAIK, you can probably stop using links in favor of the new networking stack.

**Comment by soulrebel** [comment]
I also get this all the time on 1.9.1. Plus is this a duplicate of #17611?

**Comment by pgrm** [comment]
I didn't have this issue before, but started experiencing it after setting up docker on a new server. The error shown is:

> Error response from daemon: Cannot restart container some_container_name: Cannot link to a non running container: /postgres AS /some_container_name/postgres

However, the postgres container is running and actively used by other containers without problems.

````
$ docker version
Client:
 Version:      1.11.2
 API version:  1.23
 Go version:   go1.5.4
 Git commit:   b9f10c9
 Built:        Wed Jun  1 22:00:43 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.11.2
 API version:  1.23
 Go version:   go1.5.4
 Git commit:   b9f10c9
 Built:        Wed Jun  1 22:00:43 2016
 OS/Arch:      linux/amd64 
````

````
$ docker info
Containers: 11
 Running: 11
 Paused: 0
 Stopped: 0
Images: 2313
Server Version: 1.11.2
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 2740
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge null host
Kernel Version: 4.4.0-22-generic
Operating System: Ubuntu 16.04 LTS
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 11.73 GiB
Name: rs217328
ID: A7W2:3ZT4:H2EM:55BR:F7W6:QIZ3:HOKU:6MY5:6PMA:H52X:MC6G:7NBJ
Docker Root Dir: /var/lib/docker
Debug mode (client): false
Debug mode (server): false
Username: pgrm
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
````

all containers are running with `--retstart="always"` and it hasn't been an issue before. The only way I can resolve it right now is to restart the server. Any ideas?

**Comment by talmai** [comment]
I am running node containers on my raspberrypi, and am facing the extreme opposite behavior: when the device reboots due to power outage, I see 3 or 4 different instances of the node container being restarted! The only work-around is do a 'docker rm -f XXXXX' on each running container, do a clean reboot, and then have the device functioning again.

**Comment by cpuguy83** [comment]
I'm going to close this, as this was resolved in 1.10 where docker will defer starting any containers with it's links have been started.
In 1.11.2 it was also added to wait for the daemon discovery to be ready before trying to start containers (making sure overlay networking was ready).

Any other issue should be addressed separately, and indeed is most likely a configuration error, but happy to discuss on another issue.
Thanks all.
